index.d.ts

```diff
@@ -1,10 +1,16 @@
 declare module "@iarna/toml" {
+  type AnyJson = boolean | number | string | null | JsonArray | JsonMap
+  interface JsonArray extends Array<AnyJson> {}
```

Comment:
It would be great if this matches the TOML limitation that arrays must be a single type, does this do that?

For the purposes of that, number and bigint would need to be treated as equivalent. There's no need to distinguish float from int, as `TOML.stringify` would make sure the emitted list is valid.

Reply:
No, it does not limit the array to be a single type. It may need some tricks to achieve this, I'll think about it 🤔

Reply:
I've tried the following typing (modified part only) and I'll demo the result.
Feel free to correct me if I mistake something. 😃 

```ts
type JsonArray = boolean[] | number[] | string[] | null[] | JsonMap[]
type AnyJson = boolean | number | string | null | JsonMap | JsonArray | JsonArray[]
```

----------------

Typescript compiler version
```
> tsc -v
Version 3.0.3
```

Valid usages
```ts
const arr1: JsonArray = [ 0, 1, 2 ]
const arr2: JsonArray = [ '0', '1', '2' ]
const arr3: JsonArray = [ true, false ]
const arr4: JsonArray = [ { arr: [ 1, 2, 3 ] }, { arr: [ '0', '1', '2' ] } ]
const arr5: JsonArray = [ true, false ]
const arr6: JsonArray = [ new Date(0), new Date('1979-05-27T07:32:00.999Z') ]

const map1: JsonMap = { val: 0 }
const map2: JsonMap = { val: '0' }
const map3: JsonMap = { val: true }
const map4: JsonMap = { val: [ 1, 2, 3] }
const map5: JsonMap = { val: { arr: [ 1, 2, 3 ] } }
const map6: JsonMap = { val: null }
const map7: JsonMap = { val: [ [ 1, 2, 3], [ '0', '1', '2' ], [ false, false, true ] ] }
```

Invalid usages
```ts
const arr: JsonArray = [ 0, true, '0' ]
/*
error TS2322: Type '(string | number | true)[]' is not assignable to type 'JsonArray'.
  Type '(string | number | true)[]' is not assignable to type 'string[]'.
    Type 'string | number | true' is not assignable to type 'string'.
      Type 'number' is not assignable to type 'string'.
*/
```

```ts
const map: JsonMap = {
  arr: [  0, '1' ]
}
/*
error TS2322: Type '(string | number)[]' is not assignable to type 'AnyJson'.
  Type '(string | number)[]' is not assignable to type 'string[]'.
    Type 'string | number' is not assignable to type 'string'.
      Type 'number' is not assignable to type 'string'.
*/
```
```ts
const map: JsonMap = {
  arr: [
    [ 1, 2, 3 ],
    [ 'a', 'b', 'c' ],
    [ false, '2', '3' ]
  ]
}
/*
error TS2322: Type '(number[] | (string | false)[])[]' is not assignable to type 'AnyJson'.
  Type '(number[] | (string | false)[])[]' is not assignable to type 'string[]'.
    Type 'number[] | (string | false)[]' is not assignable to type 'string'.
      Type 'number[]' is not assignable to type 'string'.
*/
```

Reply:
Is `null[]` valid for a `JsonArray`?